File: leetcode/442_find_all_duplicates_in_an_array.js
// Given an array of integers, 1 ≤ a[i] ≤ n (n = size of array), some elements appear twice and others appear once.
//
// Find all the elements that appear twice in this array.
//
// Could you do it without extra space and in O(n) runtime?
//
// Example:
// Input:
// [4,3,2,7,8,2,3,1]
//
// Output:
// [2,3]

/**
 * @param {number[]} nums
 * @return {number[]}
 */

// with extra space -> O(N)
var findDuplicates = function (nums) {
  let numbers = new Set();
  let result = [];
  for (num of nums) {
    if (!numbers.has(num)) {
      numbers.add(num);
    } else {
      result.push(num);
    }
  }
  return result;
};

// without extra space -> O(1)
var findDuplicates = function (nums) {
  let res = [];
  let index;
  let i;

  for (i = 0; i < nums.length; i++) {
    index = Math.abs(nums[i]) - 1;

    if (nums[index] < 0) res.push(index + 1);
    else nums[index] *= -1;
  }

  return res;
};

File: leetcode/442_find_all_duplicates_in_an_array.py
# Given an integer array nums of length n where all the integers of nums are in the range [1, n] and each integer appears once or twice, return an array of all the integers that appears twice.
#
# You must write an algorithm that runs in O(n) time and uses only constant extra space.
#
#
#
# Example 1:
#
# Input: nums = [4,3,2,7,8,2,3,1]
# Output: [2,3]
# Example 2:
#
# Input: nums = [1,1,2]
# Output: [1]
# Example 3:
#
# Input: nums = [1]
# Output: []
#
#
# Constraints:
#
# n == nums.length
# 1 <= n <= 105
# 1 <= nums[i] <= n
# Each element in nums appears once or twice.
# Accepted
# 307,251
# Submissions


class Solution:
    def findDuplicates(self, nums: List[int]) -> List[int]:
        output = []

        for n in nums:
            # take the absolute value since we don't know if we have visited this number yet
            m = abs(n)
            # if we haven't seen this number yet, push it into our output
            if nums[m - 1] < 0:
                output.append(m)
            else:
                # we have seen this number therefore mark it as seen
                nums[m - 1] *= -1

        return output
